### ISSUE TITLE
[RFC] vcpu_t changes

### ISSUE THREAD

**Issue by rianquinn** [Bareflank/hypervisor#851]
TL;DR

Add a #define that tells the base what vcpu_t to use, providing extensions with a way to redefine the vcpu type across all of the code.

Description

One issue that we have had for a while is that the base defines vcpu_t as bfvmm::intel_x64::vcpu which is then passed to all of the APIs throughout the base. Extensions, however, might provide their own vcpu and associated functionality and as a result, the extension has to manage two different vcpus. Although extensions can safely static_cast<> a base vcpu, its cumbersome and verbose. To solve this, I propose that we add the ability for a #define that allows an extension to define its own vcpu_t and the base would only define vcpu_t if an extension hasn't. This way nothing changes in the base, but if an extension wishes to provide its own vcpu_t it can. The #define would simply be a file path to the vcpu_t that the extension is providing. By doing this, all of the vcpu references would point to the extension's version of the vcpu and not the base.

**Comment by JaredWright** [comment]
What happens when multiple extensions ```#define vcpu_t```?

**Comment by rianquinn** [comment]
Any attempt to set vcpu_t should be wrapped with an `#ifndef`, so the leaf will be the only vcpu_t defined.

**Comment by JaredWright** [comment]
You probably won't agree with me on this one... but I think that this ```vcpu_t``` problem would go away if extensions didn't use the concept of vCPU inheritance (rather, they would "register" functionality to a vCPU provided by the base hypervisor).

But since we're nowhere near that dream of mine, this sounds like a good work around.

+1

**Comment by rianquinn** [comment]
I definitely understand the argument, but the extensions internal logic would be needlessly verbose as a result. It is possible now to implement this as we added the std::any APIs to the vCPU, but for extensions like Boxy that have a lot of internal state to manage, being able to simply do a vcpu-> or foo(vcpu, ....) is a lot cleaner. I think this approach provides a simple way to allow for both models to exist without being overly complicated or restrictive. 

With that said I would generally side with you that this is a feature that only really complicated extensions should be leveraging, and this should be our stance as a project. Unless you have a lot of state to manage, you should use the data() APIs instead of inheritance.

**Comment by rianquinn** [comment]
Closing this. This will be included in my next PR
